velox/exec/tests/utils/PlanBuilder.cpp

```diff
@@ -1014,6 +1015,287 @@ PlanBuilder& PlanBuilder::unnest(
+    const std::string& windowFunctionName,
+    const std::vector<TypePtr>& inputTypes,
+    bool nullOnFailure) {
+  if (auto signatures = exec::getWindowFunctionSignatures(windowFunctionName)) {
+    for (const auto& signature : signatures.value()) {
+      exec::SignatureBinder binder(*signature, inputTypes);
+      if (binder.tryBind()) {
+        return binder.tryResolveType(signature->returnType());
+      }
+    }
+
+    if (nullOnFailure) {
+      return nullptr;
+    }
+    throwWindowFunctionSignatureNotSupported(
+        windowFunctionName, inputTypes, signatures.value());
+  }
+
+  if (nullOnFailure) {
+    return nullptr;
+  }
+  throwWindowFunctionDoesntExist(windowFunctionName);
+  return nullptr;
+}
+
+class WindowTypeResolver {
+ public:
+  explicit WindowTypeResolver()
+      : previousHook_(core::Expressions::getResolverHook()) {
+    core::Expressions::setTypeResolverHook(
+        [&](const auto& inputs, const auto& expr, bool nullOnFailure) {
+          return resolveType(inputs, expr, nullOnFailure);
+        });
+  }
+
+  ~WindowTypeResolver() {
+    core::Expressions::setTypeResolverHook(previousHook_);
+  }
+
+  void setResultType(const TypePtr& type) {
+    resultType_ = type;
+  }
+
+ private:
+  TypePtr resolveType(
+      const std::vector<core::TypedExprPtr>& inputs,
+      const std::shared_ptr<const core::CallExpr>& expr,
+      bool nullOnFailure) const {
+    if (resultType_) {
+      return resultType_;
+    }
+
+    std::vector<TypePtr> types;
+    for (auto& input : inputs) {
+      types.push_back(input->type());
+    }
+
+    auto functionName = expr->getFunctionName();
+
+    return resolveWindowType(functionName, types, nullOnFailure);
+  }
+
+  const core::Expressions::TypeResolverHook previousHook_;
+  TypePtr resultType_;
+};
+
+const core::WindowNode::Frame createWindowFrame(
+    const duckdb::IExprWindowFrame& windowFrame,
+    const TypePtr& inputRow,
+    memory::MemoryPool* pool) {
+  core::WindowNode::Frame frame;
+  frame.type = (windowFrame.type == duckdb::WindowType::kRows)
+      ? core::WindowNode::WindowType::kRows
+      : core::WindowNode::WindowType::kRange;
+
+  auto boundTypeConversion =
+      [](duckdb::BoundType boundType) -> core::WindowNode::BoundType {
+    switch (boundType) {
+      case duckdb::BoundType::kCurrentRow:
+        return core::WindowNode::BoundType::kCurrentRow;
+      case duckdb::BoundType::kFollowing:
+        return core::WindowNode::BoundType::kFollowing;
+      case duckdb::BoundType::kPreceding:
+        return core::WindowNode::BoundType::kPreceding;
+      case duckdb::BoundType::kUnboundedFollowing:
+        return core::WindowNode::BoundType::kUnboundedFollowing;
+      case duckdb::BoundType::kUnboundedPreceding:
+        return core::WindowNode::BoundType::kUnboundedPreceding;
+    }
+    VELOX_UNREACHABLE();
+  };
+  frame.startType = boundTypeConversion(windowFrame.startType);
+  frame.startValue = windowFrame.startValue
```

Comment:
do we need to check if startValue here and endValue below are not nullptrs?

Reply:
This is done already.